src/elfpy/markets.py

```diff
@@ -1,25 +1,56 @@
 """
 Market simulators store state information when interfacing AMM pricing models with users
-
-TODO: rewrite all functions to have typed inputs
 """
 
 from dataclasses import dataclass, field
+from typing import Literal, TypeAlias
 
 import numpy as np
+from elfpy.pricing_models import ElementPricingModel, HyperdrivePricingModel, TradeResult
+from elfpy.token import TokenType
 
 import elfpy.utils.time as time_utils
-from elfpy.user import AgentWallet
 import elfpy.utils.price as price_utils
 from elfpy.utils.bcolors import Bcolors as bcolors
 from elfpy.utils.outputs import float_to_string
+from elfpy.wallet import Wallet
 
 # Currently many functions use >5 arguments.
 # These should be packaged up into shared variables, e.g.
 #     reserves = (in_reserves, out_reserves)
 #     share_prices = (init_share_price, share_price)
 # pylint: disable=too-many-arguments
 
+TradeDirection = Literal["out", "in"]
+MarketActionType: TypeAlias = Literal[
+    "close_short", "close_long", "open_short", "open_long", "add_liquidity", "remove_liquidity"
+]
+
+
+@dataclass
+class MarketAction:
+    """market action specification"""
+
+    # these two variables are required to be set by the strategy
+    action_type: MarketActionType
+    trade_amount: float
+    # wallet_address is always set automatically by the basic agent class
+    wallet_address: int
+    # mint time is set only for trades that act on existing positions (close long or close short)
+    mint_time: float = 0
+
+    def print_description_string(self) -> None:
```

Comment:
I know this func existed before the PR, but since I'm here -- is this not just an __str__ method with an extra print at the end? I think it makes more sense to overwrite __str__ with the fancy formatting, and then whoever calls `print_description_string` can instead use `print(MarketAction(...))`

Reply:
good catch, done.